spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/secretmanager/SecretManagerPropertySource.java

```diff
@@ -0,0 +1,110 @@
+
+import java.util.HashMap;
+import java.util.Map;
+
+import com.google.cloud.secretmanager.v1beta1.AccessSecretVersionResponse;
+import com.google.cloud.secretmanager.v1beta1.ProjectName;
+import com.google.cloud.secretmanager.v1beta1.Secret;
+import com.google.cloud.secretmanager.v1beta1.SecretManagerServiceClient;
+import com.google.cloud.secretmanager.v1beta1.SecretManagerServiceClient.ListSecretsPagedResponse;
+import com.google.cloud.secretmanager.v1beta1.SecretVersionName;
+
+import org.springframework.cloud.gcp.core.GcpProjectIdProvider;
+import org.springframework.core.env.EnumerablePropertySource;
+
+public class SecretManagerPropertySource extends EnumerablePropertySource<SecretManagerServiceClient> {
+
+	private static final String SECRETS_NAMESPACE = "spring-cloud-gcp.secrets.";
+
+	private static final String LATEST_VERSION_STRING = "latest";
+
+	private final Map<String, Object> properties;
+
+	private final String[] propertyNames;
+
+	public SecretManagerPropertySource(
+			String propertySourceName,
+			SecretManagerServiceClient client,
+			GcpProjectIdProvider projectIdProvider) {
+
+		super(propertySourceName, client);
+
+		Map<String, Object> propertiesMap = initializePropertiesMap(client, projectIdProvider.getProjectId());
+
+		this.properties = propertiesMap;
+		this.propertyNames = propertiesMap.keySet().toArray(new String[propertiesMap.size()]);
+	}
+
+	@Override
+	public String[] getPropertyNames() {
+		return propertyNames;
+	}
+
+	@Override
+	public Object getProperty(String name) {
+		return properties.get(name);
+	}
+
+	private static Map<String, Object> initializePropertiesMap(SecretManagerServiceClient client, String projectId) {
+		ListSecretsPagedResponse response = client.listSecrets(ProjectName.of(projectId));
+
+		HashMap<String, Object> secretsMap = new HashMap<>();
+		for (Secret secret : response.iterateAll()) {
+			String secretId = extractSecretId(secret);
+			if (secretId != null) {
+				String secretName = SECRETS_NAMESPACE + secretId;
+				String secretPayload = getSecretPayload(client, projectId, secretId);
+				secretsMap.put(secretName, secretPayload);
+			}
+		}
+
+		return secretsMap;
+	}
+
+	private static String getSecretPayload(
+			SecretManagerServiceClient client, String projectId, String secretId) {
+
+		SecretVersionName secretVersionName = SecretVersionName.newBuilder()
+				.setProject(projectId)
+				.setSecret(secretId)
+				.setSecretVersion(LATEST_VERSION_STRING)
+				.build();
+
+		AccessSecretVersionResponse response = client.accessSecretVersion(secretVersionName);
+		return response.getPayload().getData().toStringUtf8();
+	}
+
+	/**
+	 * Extracts the Secret ID from the {@link Secret}. The secret ID refers to the unique ID
+	 * given to the secret when it is saved under a GCP project.
+	 *
+	 * <p>
+	 * The secret ID is extracted from the full secret name of the form:
+	 * projects/${PROJECT_ID}/secrets/${SECRET_ID}
+	 */
+	private static String extractSecretId(Secret secret) {
```

Comment:
In general, what we've been recommending to integrators downstream is the following transformation:

```text
:pid/:sid => projects/:pid/secrets/:sid/versions/latest
:pid/:sid/:vid => projects/:pid/secrets/:sid/versions/:vid
projects/:pid/secrets/:sid => projects/:pid/secrets/:sid/versions/latest
projects/:pid/secrets/:sid/versions/:vid => projects/:pid/secrets/:sid/versions/:vid
```

That way, the result of the parsing is always a fully-qualified secret version reference which you can pass directly to AccessSecretVersion without needing to complicate that method.

Reply:
Noted, I simplified the method a bit more, but the `.split()` I am doing is somewhat inevitable because I need to get the shortened secret name which is used as the Key of the Spring environment property.